crates/evm-ethereum/src/execute.rs

```diff
@@ -0,0 +1,826 @@
+    /// Chain specific evm config
+    evm_spec: EthEvmExecutor<EvmConfig>,
+    /// The state to use for execution
+    state: State<DB>,
+    /// Optional inspector stack for debugging
+    inspector: Option<InspectorStack>,
+}
+
+impl<EvmConfig, DB> EthBlockExecutor<EvmConfig, DB> {
+    /// Creates a new Ethereum block executor.
+    pub fn new(chain_spec: Arc<ChainSpec>, evm_config: EvmConfig, state: State<DB>) -> Self {
+        Self { evm_spec: EthEvmExecutor { chain_spec, evm_config }, state, inspector: None }
+    }
+
+    /// Sets the inspector stack for debugging.
+    pub fn with_inspector(mut self, inspector: Option<InspectorStack>) -> Self {
+        self.inspector = inspector;
+        self
+    }
+
+    #[inline]
+    fn chain_spec(&self) -> &ChainSpec {
+        &self.evm_spec.chain_spec
+    }
+
+    /// Returns mutable reference to the state that wraps the underlying database.
+    #[allow(unused)]
+    fn state_mut(&mut self) -> &mut State<DB> {
+        &mut self.state
+    }
+}
+
+impl<EvmConfig, DB> EthBlockExecutor<EvmConfig, DB>
+where
+    EvmConfig: ConfigureEvm,
+    // TODO(mattsse): get rid of this
+    EvmConfig: ConfigureEvmEnv<TxMeta = ()>,
+    DB: Database<Error = ProviderError>,
+{
+    /// Configures a new evm configuration and block environment for the given block.
+    ///
+    /// # Caution
+    ///
+    /// This does not initialize the tx environment.
+    fn evm_env_for_block(&self, header: &Header, total_difficulty: U256) -> EnvWithHandlerCfg {
+        let mut cfg = CfgEnvWithHandlerCfg::new(Default::default(), Default::default());
+        let mut block_env = BlockEnv::default();
+        EvmConfig::fill_cfg_and_block_env(
+            &mut cfg,
+            &mut block_env,
+            self.chain_spec(),
+            header,
+            total_difficulty,
+        );
+
+        EnvWithHandlerCfg::new_with_cfg_env(cfg, block_env, Default::default())
+    }
+
+    /// Execute a single block and apply the state changes to the internal state.
+    ///
+    /// Returns the receipts of the transactions in the block and the total gas used.
+    ///
+    /// Returns an error if execution fails or receipt verification fails.
+    fn execute_and_verify(
+        &mut self,
+        block: &BlockWithSenders,
+        total_difficulty: U256,
+    ) -> Result<(Vec<Receipt>, u64), BlockExecutionError> {
+        // 1. prepare state on new block
+        self.on_new_block(&block.header);
+
+        // 2. configure the evm and execute
+        let env = self.evm_env_for_block(&block.header, total_difficulty);
+
+        let (receipts, gas_used) = {
+            if let Some(inspector) = self.inspector.as_mut() {
+                let evm = self.evm_spec.evm_config.evm_with_env_and_inspector(
+                    &mut self.state,
+                    env,
+                    inspector,
+                );
+                self.evm_spec.execute_pre_and_transactions(block, evm)?
```

Comment:
this just makes it more verbose because I always have to pass the evm around.

for now this is fine but all of this will be abstracted into execution stages, pre,txs,post